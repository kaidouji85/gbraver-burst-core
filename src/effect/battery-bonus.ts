/**
 * バッテリーによるダメージボーナス
 * @param attackBattery 攻撃側が出したバッテリー
 * @param defenseBattery 防御側が出したバッテリー
 * @returns ダメージボーナス
 */
export function batteryBonus(
  attackBattery: number,
  defenseBattery: number,
): number {
  return 100 * Math.max(attackBattery - defenseBattery - 1, 0);
}
