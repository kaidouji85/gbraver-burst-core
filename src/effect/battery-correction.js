// @flow

import type {ArmdozerEffect} from "../state/armdozer-effect";
import type {BatteryCommand} from "../command/battery";

/**
 * 補正後のバッテリー宣言を計算する
 *
 * @param command バッテリーコマンド
 * @param effects アームドーザ効果
 * @return 補正後のバッテリー
 */
export function correctedBattery(command: BatteryCommand, effects: ArmdozerEffect[]): number {
  if (hasIgnoreBatteryCorrection(effects)) {
    return command.battery;
  }

  const totalCorrection = totalBatteryCorrection(effects);
  const corrected = command.battery + totalCorrection;
  return Math.max(corrected, 0);
}

/**
 * バッテリー補正無効の効果が適用されているか否かを判定する
 * trueでバッテリー補正無効である
 *
 * @param effects アームドーザエフェクト
 * @return 判定結果
 */
export function hasIgnoreBatteryCorrection(effects: ArmdozerEffect[]): boolean {
  return effects
    .filter(v => v.type === 'IgnoreBatteryCorrection')
    .length > 0;
}

/**
 * アームドーザ効果のバッテリー補正合計値を計算する
 *
 * @param effects アームドーザエフェクト
 * @return バッテリー補正値合計
 */
export function totalBatteryCorrection(effects: ArmdozerEffect[]): number {
  return effects
    .map(v => (v.type === 'BatteryCorrection') ? v.batteryCorrection : 0)
    .reduce((a, b) => a + b, 0);
}