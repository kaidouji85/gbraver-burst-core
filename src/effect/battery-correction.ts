import type { BatteryCommand } from "../command/battery";
import type { ArmdozerEffect } from "../state/armdozer-effect";

/**
 * アームドーザ効果のバッテリー補正合計値を計算する
 *
 * @param effects アームドーザエフェクト
 * @return バッテリー補正値合計
 */
export function totalBatteryCorrection(effects: ArmdozerEffect[]): number {
  return effects
    .map((v) => (v.type === "BatteryCorrection" ? v.batteryCorrection : 0))
    .reduce((a, b) => a + b, 0);
}

/**
 * 補正後のバッテリー宣言を計算する
 *
 * @param command バッテリーコマンド
 * @param effects アームドーザ効果
 * @return 補正後のバッテリー
 */
export function correctedBattery(
  command: BatteryCommand,
  effects: ArmdozerEffect[]
): number {
  if (command.battery <= 0) {
    return command.battery;
  }

  const totalCorrection = totalBatteryCorrection(effects);
  const corrected = command.battery + totalCorrection;
  return Math.max(corrected, 0);
}
