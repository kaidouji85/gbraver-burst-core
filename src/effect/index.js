// @flow

import type {InputCommand} from "./input-command/input-command";
import type {Battle} from "./battle/battle";
import type {TurnChange} from "./turn-change/turn-change";
import type {StartGame} from "./start-game/start-game";
import type {BurstEffect} from "./burst/burst-effect";
import type {GameEnd} from "./game-end/game-end";
import type {BatteryDeclaration} from "./battery-declaration/battery-declaration";
import type {Reflect} from "./reflect/reflect";
import type {UpdateRemainingTurn} from "./update-remaning-turn/update-remaining-turn";

/** ゲーム中に発生する効果をまとめたもの */
export type Effect =
  StartGame |
  GameEnd |
  InputCommand |
  Reflect |
  BatteryDeclaration |
  Battle |
  TurnChange |
  BurstEffect |
  UpdateRemainingTurn;
