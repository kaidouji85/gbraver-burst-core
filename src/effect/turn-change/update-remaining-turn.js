// @flow

import type {ArmdozerEffect} from "../../game-state/armdozer/armdozer-effect";

/**
 * アームドーザ効果の継続ターンを更新する
 * 残りターンが0になったものはリストから外す
 *
 * @param effects 更新前のアームドーザ効果
 * @return 更新結果
 */
export function updateRemainingTurn(effects: ArmdozerEffect[]): ArmdozerEffect[] {
  return effects
    .map(v => isFinite(v.remainingTurn)
      ? {...v, remainingTurn: v.remainingTurn - 1}
      : v
    ).filter(v => (0 < v.remainingTurn));
}
