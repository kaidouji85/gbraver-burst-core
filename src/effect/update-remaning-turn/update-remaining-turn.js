// @flow

import type {ArmdozerEffect, PlayerId} from "../..";

/** 終了したアームドーザ効果 */
export type EndArmdozerEffect = {
  /** プレイヤーID */
  playerId: PlayerId,
  /** 終了したエフェクト */
  effect: ArmdozerEffect
};

/**
 * 効果継続ターン数を更新する
 */
export type UpdateRemainingTurn = {
  name: 'UpdateRemainingTurn',

  /** 終了したアームドーザ効果 */
  endArmdozerEffects: EndArmdozerEffect[],
};