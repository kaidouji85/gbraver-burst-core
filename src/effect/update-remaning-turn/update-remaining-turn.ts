import { z } from "zod";

import { PlayerId,PlayerIdSchema } from "../../player/player";
import { ArmdozerEffect, ArmdozerEffectSchema } from "../../state/armdozer-effect";

/** 終了したアームドーザ効果 */
export type EndArmdozerEffect = Readonly<{
  /** プレイヤーID */
  playerId: PlayerId;
  /** 終了したエフェクト */
  effect: ArmdozerEffect;
}>;

/** EndArmdozerEffect zodスキーマ */
export const EndArmdozerEffectSchema = z.object({
  playerId: PlayerIdSchema,
  effect: ArmdozerEffectSchema,
});

/** 効果継続ターン数を更新する */
export type UpdateRemainingTurn = Readonly<{
  name: "UpdateRemainingTurn";
  /** 終了したアームドーザ効果 */
  endArmdozerEffects: EndArmdozerEffect[];
}>;

/** UpdateRemainingTurn zodスキーマ */
export const UpdateRemainingTurnSchema = z.object({
  name: z.literal("UpdateRemainingTurn"),
  endArmdozerEffects: z.array(EndArmdozerEffectSchema),
});