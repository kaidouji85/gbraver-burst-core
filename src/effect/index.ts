import { z } from "zod";

import { BatteryDeclaration, BatteryDeclarationSchema } from "./battery-declaration/battery-declaration";
import { Battle, BattleSchema } from "./battle/battle";
import { BurstEffect, BurstEffectSchema } from "./burst/burst-effect";
import { GameEnd, GameEndSchema } from "./game-end/game-end";
import { InputCommand, InputCommandSchema } from "./input-command/input-command";
import { PilotSkillEffect, PilotSkillEffectSchema } from "./pilot-skill/pilot-skill-effect";
import { Reflect, ReflectSchema } from "./reflect/reflect";
import { RightItself, RightItselfSchema } from "./right-itself/right-itself";
import { StartGame, StartGameSchema } from "./start-game/start-game";
import { TurnChange, TurnChangeSchema } from "./turn-change/turn-change";
import { UpdateRemainingTurn, UpdateRemainingTurnSchema } from "./update-remaning-turn/update-remaining-turn";

/** ゲーム中に発生する効果をまとめたもの */
export type Effect =
  | StartGame
  | GameEnd
  | InputCommand
  | Reflect
  | BatteryDeclaration
  | Battle
  | TurnChange
  | BurstEffect
  | UpdateRemainingTurn
  | RightItself
  | PilotSkillEffect;

/** Effect zodスキーマ */
export const EffectSchema = z.union([
  StartGameSchema,
  GameEndSchema,
  InputCommandSchema,
  ReflectSchema,
  BatteryDeclarationSchema,
  BattleSchema,
  TurnChangeSchema,
  BurstEffectSchema,
  UpdateRemainingTurnSchema,
  RightItselfSchema,
  PilotSkillEffectSchema,
]);