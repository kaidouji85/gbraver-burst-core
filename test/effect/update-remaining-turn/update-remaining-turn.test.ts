import { updateRemainingTurn } from "../../../src/effect/update-remaining-turn";
import { lastState } from "./update-remaining-turn.data";

test("効果継続ターン更新が正しく処理される", () => {
  const result = updateRemainingTurn(lastState);
  expect(result).toMatchSnapshot();
});
