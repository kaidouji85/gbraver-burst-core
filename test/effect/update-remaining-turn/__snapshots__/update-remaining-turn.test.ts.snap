// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`効果継続ターン更新が正しく処理される 1`] = `
{
  "activePlayerId": "",
  "effect": {
    "endArmdozerEffects": [
      {
        "effect": {
          "period": {
            "remainingTurn": 0,
            "type": "TurnLimit",
          },
          "type": "Empty",
        },
        "playerId": "player1",
      },
      {
        "effect": {
          "period": {
            "remainingTurn": 0,
            "type": "TurnLimit",
          },
          "type": "Empty",
        },
        "playerId": "player2",
      },
    ],
    "name": "UpdateRemainingTurn",
  },
  "players": [
    {
      "armdozer": {
        "battery": 5,
        "batteryAutoRecovery": 3,
        "burst": {
          "recoverBattery": 5,
          "turnStartBatteryCorrect": 1,
          "type": "RecoverBattery",
        },
        "effects": [
          {
            "period": {
              "type": "SpecialPeriod",
            },
            "type": "Empty",
          },
          {
            "period": {
              "remainingTurn": 3,
              "type": "TurnLimit",
            },
            "type": "Empty",
          },
        ],
        "enableBurst": true,
        "hp": 3000,
        "id": "id",
        "maxBattery": 5,
        "maxHp": 3000,
        "name": "name",
        "power": 2000,
        "speed": 2000,
      },
      "pilot": {
        "enableSkill": true,
        "id": "EMPTY_PLAYER",
        "name": "名無し",
        "skill": {
          "recoverBattery": 2,
          "type": "RecoverBatterySkill",
        },
      },
      "playerId": "player1",
    },
    {
      "armdozer": {
        "battery": 5,
        "batteryAutoRecovery": 3,
        "burst": {
          "recoverBattery": 5,
          "turnStartBatteryCorrect": 1,
          "type": "RecoverBattery",
        },
        "effects": [
          {
            "period": {
              "remainingTurn": 2,
              "type": "TurnLimit",
            },
            "type": "Empty",
          },
        ],
        "enableBurst": true,
        "hp": 3000,
        "id": "id",
        "maxBattery": 5,
        "maxHp": 3000,
        "name": "name",
        "power": 2000,
        "speed": 2000,
      },
      "pilot": {
        "enableSkill": true,
        "id": "EMPTY_PLAYER",
        "name": "名無し",
        "skill": {
          "recoverBattery": 2,
          "type": "RecoverBatterySkill",
        },
      },
      "playerId": "player2",
    },
  ],
}
`;
